### ISSUE TITLE
Bug in version 1.0.32 onwards.

### ISSUE THREAD

**Issue by byrning** [Bilal-S/iis2tomcat#47]
Hi,

We were running version 1.0.30 without any issues. We decided to upgrade to 1.0.35 and did so by simply removing and then placing the new boncode dll's into GAC. Errors pretty much immediately started occurring.

The error in question is:

```
2017-03-12 16:39:28 1.0.35 ERROR 

Unable to read data from the transport connection: An established connection was aborted by the software in your host machine.
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ComunicateWithTomcat()
2017-03-12 16:39:28 failed Server Connection was aborted
2017-03-12 16:39:28 1.0.35 ERROR 
TCP Client level -- Server/Port:127.0.0.1/8009
Connection between Tomcat and IIS experienced error. If you restarted Tomcat this is expected. failed:Server Connection was aborted
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ConnectionError(String message, String messageType)
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ComunicateWithTomcat()
   at BonCodeAJP13.BonCodeAJP13ServerConnection.HandleConnection()
   at BonCodeAJP13.BonCodeAJP13ServerConnection.p_CreateConnection(BonCodeAJP13PacketCollection packetsToSend)
```

In an attempt to try and locate the bug, we went back to 1.0.30 and then upgraded to each new version after that. 1.0.31 runs fine, but the errors start again in 1.0.32, so I'm thinking that the bug was introduced there.    

The server is:
Windows 2008R2, IIS 7.5 running a couple hundred websites, with boncode installed in GAC.

The config file in c:\windows is as follows:

```
<Settings>
<Server>127.0.0.1</Server>
<Port>8009</Port>
<EnableRemoteAdmin>False</EnableRemoteAdmin>
<EnableHeaderDataSupport>True</EnableHeaderDataSupport>
<ForceSecureSession>False</ForceSecureSession>
<AllowEmptyHeaders>False</AllowEmptyHeaders>
<MaxConnections>1000</MaxConnections>
<PacketSize>65536</PacketSize>
<LogDir>c:\inetpub\logs\boncode\</LogDir>
<LogLevel>1</LogLevel>
</Settings>
```
Let me know if you need anything else.

Thanks.

**Comment by Bilal-S** [comment]
Thanks for letting me know. I am assuming you restarted tomcat and IIS.
I will see whether I can do a test case later this week.

**Comment by Bilal-S** [comment]
Did you enable windows event log recording? If so, can you share anything in the log records?

**Comment by Bilal-S** [comment]
If you run 1.0.35 with this setting do you still see the error
`<MaxConnections>0</MaxConnections>`

**Comment by byrning** [comment]
Yes, restarted between versions.  The event log error details are pretty much the same as the log file details. See below:

```
- <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="BonCodeConnector" /> 
  <EventID Qualifiers="0">417</EventID> 
  <Level>2</Level> 
  <Task>0</Task> 
  <Keywords>0x80000000000000</Keywords> 
  <TimeCreated SystemTime="2017-03-22T03:48:34.000000000Z" /> 
  <EventRecordID>163047</EventRecordID> 
  <Channel>Application</Channel> 
  <Computer>web1</Computer> 
  <Security /> 
  </System>
- <EventData>
  <Data>Connection error 2: Connection between Tomcat and IIS experienced error. If you restarted Tomcat this is expected. at BonCodeAJP13.BonCodeAJP13ServerConnection.ConnectionError() at BonCodeAJP13.BonCodeAJP13ServerConnection.p_CreateConnection(BonCodeAJP13PacketCollection packetsToSend) at BonCodeIIS.BonCodeCallHandler.ProcessRequest(HttpContext context)</Data> 
  </EventData>
  </Event>
```

Running 1.0.35 as suggested with MaxConnections = 0 appears to resolve the issue - so far so good - thanks.  

But, everything runs completely error free with MaxConnections = 1000 and version <= 1.0.31, so I'm still thinking something isn't right with the connection changes made in 1.0.32. The system also isn't stressed when the errors start occurring, so it's not like it's getting anywhere close to the 1000 limit - also this is within a minute or 2 of starting up the app pool & boncode.

Thanks

**Comment by Bilal-S** [comment]
Thanks for the feedback. 
Do you have a minimal test case that you could share with me?

**Comment by byrning** [comment]
Sure, I've tested it on 2 servers.

Windows Server 2016 Standard  or  Windows Server 2008R2 Standard
Java: 1.8.0_121 (Oracle Corporation) 64bit 
Tomcat: Apache Tomcat/8.0.41
Lucee: Lucee 5.1.3.4-SNAPSHOT
Problem Boncode Versions: v1.0.32 - v1.0.35 installed in GAC.

Boncode Settings file in c:\windows\:
```
<Settings>
<Server>127.0.0.1</Server>
<Port>8009</Port>
<EnableRemoteAdmin>False</EnableRemoteAdmin>
<EnableHeaderDataSupport>True</EnableHeaderDataSupport>
<ForceSecureSession>False</ForceSecureSession>
<AllowEmptyHeaders>False</AllowEmptyHeaders>
<MaxConnections>1000</MaxConnections>
<PacketSize>65536</PacketSize>
<LogDir>c:\inetpub\logs\boncode\</LogDir>
<LogLevel>1</LogLevel>
</Settings>
```

Tomcat AJP connector in server.xml:
`<Connector port="8009" protocol="AJP/1.3" redirectPort="8443" packetSize="65536" maxThreads="1000" connectionTimeout="60000" address="127.0.0.1" />`

2 websites setup in IIS: 
 www.test1.com 
 www.test2.com
 
Test 1 and Test 2 simply have one index.cfm file each, with code:
`<cfoutput>#1+1#</cfoutput>`

Hit both of those a couple times with a benchmarking program such as "ab -n 1000 -c 10 http://www.test1.com/" and you'll start to see the following error show up in the logs. 

```
Unable to read data from the transport connection: An established connection was aborted by the software in your host machine.
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ComunicateWithTomcat()
2017-03-25 22:14:32 failed Server Connection was aborted
2017-03-25 22:14:32 1.0.35 ERROR 
TCP Client level -- Server/Port:127.0.0.1/8009
Connection between Tomcat and IIS experienced error. If you restarted Tomcat this is expected. failed:Server Connection was aborted
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ConnectionError(String message, String messageType)
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ComunicateWithTomcat()
   at BonCodeAJP13.BonCodeAJP13ServerConnection.HandleConnection()
   at BonCodeAJP13.BonCodeAJP13ServerConnection.p_CreateConnection(BonCodeAJP13PacketCollection packetsToSend)
2017-03-25 22:14:33 1.0.35 ERROR 
```

Setting **MaxConnections** to **0** makes the problem go away, but everything worked perfectly without having to do that in versions <= 1.0.31.

Thanks.

**Comment by Bilal-S** [comment]
OK. Thanks for the details. I will replicate.

**Comment by Bilal-S** [comment]
@byrning 
Can you validate the behavior for me when you change one of the settings.
Can you set `<ReadTimeOut>1500000</ReadTimeOut>` and `<WriteTimeout>300000</WriteTimeOut>`  and let me know whether you see the behavior reoccur after that even when setting a value for `<MaxConnections>`
Thanks,
Bilal

**Comment by Bilal-S** [comment]
Let me know if version 1.0.36 fixes this issue.

**Comment by byrning** [comment]
Apologies for the delay in response, I've been buried.

> Can you set <ReadTimeOut>1500000</ReadTimeOut> and <WriteTimeout>300000</WriteTimeOut> and let me know whether you see the behavior reoccur after that even when setting a value for <MaxConnections>

So I ran a couple tests with those 2 settings and while I did not see the errors reappear, there was a MASSIVE slowdown with the requests - see output below.

**Before settings:**
Concurrency Level:      10
Time taken for tests:   11.170 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      2286000 bytes
HTML transferred:       1843000 bytes
Requests per second:    89.52 [#/sec] (mean)
Time per request:       111.705 [ms] (mean)
Time per request:       11.170 [ms] (mean, across all concurrent requests)
Transfer rate:          199.85 [Kbytes/sec] received

**After Settings:** 
Concurrency Level:      10
Time taken for tests:   136.172 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      2286000 bytes
HTML transferred:       1843000 bytes
Requests per second:    7.34 [#/sec] (mean)
Time per request:       1361.722 [ms] (mean)
Time per request:       136.172 [ms] (mean, across all concurrent requests)
Transfer rate:          16.39 [Kbytes/sec] received
 
As you can see, requests are about 13x slower with those 2 settings and with several tests ab would completely timeout, so I wouldn't recommend those settings.


> Let me know if version 1.0.36 fixes this issue.

Unfortunately the error still occurs with 1.0.36B.

```
Unable to read data from the transport connection: An established connection was aborted by the software in your host machine.
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ComunicateWithTomcat()
2017-05-06 19:37:51 failed Server Connection was aborted
2017-05-06 19:37:51 1.0.36 ERROR 
TCP Client level -- Server/Port:127.0.0.1/8009
Connection between Tomcat and IIS experienced error. If you restarted Tomcat this is expected. failed:Server Connection was aborted
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ConnectionError(String message, String messageType)
   at BonCodeAJP13.BonCodeAJP13ServerConnection.ComunicateWithTomcat()
   at BonCodeAJP13.BonCodeAJP13ServerConnection.HandleConnection()
   at BonCodeAJP13.BonCodeAJP13ServerConnection.p_CreateConnection(BonCodeAJP13PacketCollection packetsToSend)
```

We've been running 1.0.35 in production with MaxConnections=0 for a while now and that works well. I wouldn't recommend the MaxConnections > 0 setting for versions 1.0.32 onwards.

Thanks.

**Comment by byrning** [comment]
**So I think I've tracked down the issue.**

1.0.32 was the version that you fixed connection pooling, so pooling hasn't worked before that right? I had a look at your 1.0.32 code changes and it all looked good, so figured the issue must be elsewhere, yet still related.  I did a bit of poking around and you'll notice that in the test case above, our tomcat **<Connector** string has **connectionTimeout="60000"**. In the "Configuring the Connector for Performance" section of your docs, you mentioned setting the **connectionTimeout** slightly higher than the **IIS idle-timeout** - your example was 121000sec for tomcat and 2min for IIS. 

Our IIS idle-timeout is 120**min**, or on a new box or pool the default is 20min, but I tried setting it to 2min - even though that didn't seem right to me because the IIS timeout is for the entire worker process, not threads or connections it starts up - and hit it with ab.. still got the error. 

I figured the issue must be related to timeouts, that tomcat was closing the connection before IIS closed it. The IIS idle-timeout doesn't fix things, because it's for the entire worker process, so it remains active as long as it's doing something, whereas tomcat is still going to close the connection when it hits 121 seconds.

The solution to all of this is **keepAliveTimeout**. The setting will ensure that tomcat does not close the connection, it waits for the client to close it. I've been hitting the server with ab for 30min now and no errors, so when using maxconnection > 0, one should have the following:

**Tomcat connector needs:**
connectionTimeout="60000" keepAliveTimeout="-1"

**IIS idle-timeout**
20min or greater, we have it at 120min. I don't recommend setting it to 2min as outlined in your docs, because one really doesn't want the worker process recycling if it really doesn't need to.
